docs/development/workflow/development_workflow.rst

```diff
@@ -149,6 +153,45 @@ document:
   ``astropy``; in prior versions of this documentation it was referred to as
   ``upstream``.
 
+
+.. _pre-commit:
+
+Pre-commit
+**********
+
+All of the coding style checks described in :ref:`code-style` can be performed automatically
+when you make a git commit using our provided `pre-commit hook <https://pre-commit.com/>`_
+for git, for more information see
+`git hooks <https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks#_git_hooks>`_.
+We encourage you to setup and use these hooks to ensure that your code always meets
+our coding style standards. This can be done by installing ``pre-commit`` in the root
+of your astropy repository by running::
+
+    pip install pre-commit
+    pre-commit install
+
+For more detailed instructions on installing ``pre-commit``, see the
+`install guide <https://pre-commit.com/#install>`_. Once this installation is
+complete, all the coding style checks will be run each time you commit and the
+necessary changes will automatically be applied to your code if possible.
+
+.. note::
+  The changes made by ``pre-commit`` will not be automatically staged, so you
+  will need to review and re-stage any files that ``pre-commit`` has changed.
+
+In general, git will not allow you to commit until the ``pre-commit`` hook has
+run successfully. If you need to make a commit which fails the ``pre-commit`` checks,
+you can skip (strongly discouraged) these checks by running::
+
+  git commit --no-verify
+
+Ff you do not want to use ``pre-commit`` as part of your git workflow, you can
+still run the checks manually (see, :ref:`code-style`) using::
+
+  tox -e codestyle
+
+Again, this will automatically apply the necessary changes to your code if possible.
```

Comment:
This is surprising to me since I thought tox is generally working in a sandboxed environment so I wouldn't expect it to have a side-effect of touching the code.  Maybe this like how `build_docs` will build in the source docs tree (which I also found surprising)?

Reply:
I was surprised by this too, it was brought up in https://github.com/astropy/astropy/pull/13552#discussion_r945782301. I specifically tested it and found that it did change the code.

Reply:
The codestyle environment uses different tox settings (specifically `skip_install = true`) than the test environments, but all tox envs have write access to the git checkout from where they are run, just the test envs make a tarball and install and explicitly test against that rather than make use of the checkout. (by choice, both are possible)